specs/pluggable_alert_delivery.adoc

```diff
@@ -0,0 +1,372 @@
+= Pluggable Alerts Delivery End-Points
+
+== Introduction
+
+The alerting module in tendrl needs to alert/notify
+
+* State changes of tendrl managed entities
+* Threshold breaches of utilizations of tendrl managed entities.
+
+via mediums such as:
+
+* SNMP-traps
+* email
+
+The different sources of alerts are:
+
+* Performance monitoring threshold breach alerts generated by collectd.
+* Alerts related to state changes of tendrl managed sds specific entities
+  generated by sds-integration.
+  ex: Tendrl/gluster-integration, Tendrl/ceph-integration
+
+== Problem description
+
+* The different tendrl supported alerting means need to be pluggable in the
+  sense that adding any new alert notifying means should be an easy task and
+  should not involve much changes in the core alerting framework.
+* The alerting mechanism should have a way to set importance of alerts so that
+  the alerts desingated as important will be notified to users even if the
+  alerting module crashed and hence the alert could not be notified but will
+  be alerted once the alerting module is started back.
+
+== Use Cases
+
+* Dynamic addition of new means of notifying an alert.
+
+== Proposed change
+
+* The alerts raised are persisted in '/alerts' directory of etcd before sending
+  out notifications. The flow is as below:
+  ** One socket is connected to for writes from collectd, read-only from the
```

Comment:
I have asked @anmolbabu to use Co-op Socket (http://www.gevent.org/gevent.socket.html) which is part of our consumption of green threads (via gevent library in python)

Please dont use normal processes/threads/sockets for any Tendrl functionality written in python,  Use the gevent library (http://www.gevent.org/)